contracts/src/mixins/QuotaMixin.sol

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: GNU GPLv3
+pragma solidity ^0.8.29;
+
+import {
+    SafeERC20
+} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "../interfaces/QuotaManager.sol";
+import "../libs/CoreLib.sol";
+
+struct QuotaStats {
+    uint128 tokenBalance; // uint96 might also be enough, as this would be more tokens than most project have in circulation
+    uint64 usedQuota; // Quota is limited to 64 bits (current Quota == Bytes of data, so that should be ok)
+    uint64 nextQuotaReset; // timestamps are Safe to limit to 64 bits
+}
+
+abstract contract QuotaMixin is IQuotaManager {
+    using SafeERC20 for IERC20;
+
+    event Withdraw(address indexed signer, uint256 indexed amount);
+    event Deposit(address indexed signer, uint256 indexed amount);
+
+    struct QuotaMixinConfig {
+        uint64 timeframeQuotaReset;
+        uint16 requiredQuotaMultiplier;
+        uint32 freeQuotaPerDepositedFeeToken;
+        uint32 maxFreeQuota;
+        address feeToken;
+        uint8 feeTokenDecimals;
+    }
```

Comment:
Since this only matters for creation code, I'm not sure I understand why the smaller integer size are needed. All immutable (AFAIU) are `PUSH32`s anyway, so you don't really save anything and instead just generate a bunch of masking code.

I would also remove `feeTokenDecimals` and instead just make `feeQuotaPerDepositedFeeToken` not be in base units and be in atoms instead. This simplifies the runtime code (no `10 ** decimals`).